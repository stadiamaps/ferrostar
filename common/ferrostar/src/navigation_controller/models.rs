//! State and configuration data models.

use crate::deviation_detection::{RouteDeviation, RouteDeviationTracking};
use crate::models::{RouteStep, SpokenInstruction, UserLocation, VisualInstruction, Waypoint};
#[cfg(feature = "alloc")]
use alloc::vec::Vec;
use geo::LineString;
#[cfg(feature = "wasm-bindgen")]
use serde::{Deserialize, Serialize};

/// High-level state describing progress through a route.
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(feature = "wasm-bindgen", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
pub struct TripProgress {
    /// The distance to the next maneuver, in meters.
    pub distance_to_next_maneuver: f64,
    /// The total distance remaining in the trip, in meters.
    ///
    /// This is the sum of the distance remaining in the current step and the distance remaining in all subsequent steps.
    pub distance_remaining: f64,
    /// The total duration remaining in the trip, in seconds.
    pub duration_remaining: f64,
}

/// The state of a navigation session.
///
/// This is produced by [`NavigationController`](super::NavigationController) methods
/// including [`get_initial_state`](super::NavigationController::get_initial_state)
/// and [`update_user_location`](super::NavigationController::update_user_location).
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Serialize, Deserialize))]
pub enum TripState {
    /// The navigation controller is idle and there is no active trip.
    Idle,
    #[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
    /// The navigation controller is actively navigating a trip.
    Navigating {
        /// The index of the closest coordinate to the user's snapped location.
        ///
        /// This index is relative to the *current* [`RouteStep`]'s geometry.
        current_step_geometry_index: Option<u64>,
        /// A location on the line string that
        snapped_user_location: UserLocation,
        /// The ordered list of steps that remain in the trip.
        ///
        /// The step at the front of the list is always the current step.
        /// We currently assume that you cannot move backward to a previous step.
        remaining_steps: Vec<RouteStep>,
        /// Remaining waypoints to visit on the route.
        ///
        /// The waypoint at the front of the list is always the *next* waypoint "goal."
        /// Unlike the current step, there is no value in tracking the "current" waypoint,
        /// as the main use of waypoints is recalculation when the user deviates from the route.
        /// (In most use cases, a route will have only two waypoints, but more complex use cases
        /// may have multiple intervening points that are visited along the route.)
        /// This list is updated as the user advances through the route.
        remaining_waypoints: Vec<Waypoint>,
        /// The trip progress includes information that is useful for showing the
        /// user's progress along the full navigation trip, the route and its components.
        progress: TripProgress,
        /// The route deviation status: is the user following the route or not?
        deviation: RouteDeviation,
        /// The visual instruction that should be displayed in the user interface.
        visual_instruction: Option<VisualInstruction>,
        /// The most recent spoken instruction that should be synthesized using TTS.
        ///
        /// Note it is the responsibility of the platform layer to ensure that utterances are not synthesized multiple times. This property simply reports the current spoken instruction.
        spoken_instruction: Option<SpokenInstruction>,
        /// Annotation data at the current location.
        /// This is represented as a json formatted byte array to allow for flexible encoding of custom annotations.
        annotation_json: Option<String>,
    },
    /// The navigation controller has reached the end of the trip.
    Complete,
}

pub enum StepAdvanceStatus {
    /// Navigation has advanced, and the information on the next step is embedded.
    Advanced {
        step: RouteStep,
        linestring: LineString,
    },
    /// Navigation has reached the end of the route.
    EndOfRoute,
}

/// Controls filtering/post-processing of user course by the [`NavigationController`].
#[derive(Debug, Copy, Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Deserialize))]
pub enum CourseFiltering {
    /// Snap the user's course to the current step's linestring using the next index in the step's geometry.
    ///
    // TODO: We could make this more flexible by allowing the user to specify number of indices to average, etc.
    SnapToRoute,

    /// Use the raw course as reported by the location provider with no processing.
    Raw,
}

/// The step advance mode describes when the current maneuver has been successfully completed,
/// and we should advance to the next step.
#[derive(Debug, Copy, Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Deserialize))]
pub enum StepAdvanceMode {
    /// Never advances to the next step automatically;
    /// requires calling [`NavigationController::advance_to_next_step`](super::NavigationController::advance_to_next_step).
    ///
    /// You can use this to implement custom behaviors in external code.
    Manual,
    /// Automatically advances when the user's location is close enough to the end of the step
    #[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
    DistanceToEndOfStep {
        /// Distance to the last waypoint in the step, measured in meters, at which to advance.
        distance: u16,
        /// The minimum required horizontal accuracy of the user location, in meters.
        /// Values larger than this cannot trigger a step advance.
        minimum_horizontal_accuracy: u16,
    },
    /// Automatically advances when the user's distance to the *next* step's linestring  is less
    /// than the distance to the current step's linestring.
    #[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
    RelativeLineStringDistance {
        /// The minimum required horizontal accuracy of the user location, in meters.
        /// Values larger than this cannot trigger a step advance.
        minimum_horizontal_accuracy: u16,
        /// At this (optional) distance, navigation should advance to the next step regardless
        /// of which `LineString` appears closer.
        automatic_advance_distance: Option<u16>,
    },
}

#[derive(Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(feature = "wasm-bindgen", derive(Deserialize))]
#[cfg_attr(feature = "wasm-bindgen", serde(rename_all = "camelCase"))]
pub struct NavigationControllerConfig {
    /// Configures when navigation advances to the next step in the route.
    pub step_advance: StepAdvanceMode,
    /// Configures when the user is deemed to be off course.
    ///
    /// NOTE: This is distinct from the action that is taken.
    /// It is only the determination that the user has deviated from the expected route.
    pub route_deviation_tracking: RouteDeviationTracking,
    /// Configures how the heading component of the snapped location is reported in [`TripState`].
    pub snapped_location_course_filtering: CourseFiltering,
}
