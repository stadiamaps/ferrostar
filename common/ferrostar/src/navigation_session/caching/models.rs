use serde::{Deserialize, Serialize};

#[cfg(all(feature = "uniffi", not(feature = "wasm_js")))]
use crate::{models::Route, navigation_controller::models::TripState, UtcDateTime};

#[cfg(all(feature = "uniffi", not(feature = "wasm_js")))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
pub struct NavigationCachingConfig {
    pub cache_interval_seconds: i64,
    pub max_age_seconds: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg(all(feature = "uniffi", not(feature = "wasm_js")))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
pub struct NavigationSessionSnapshot {
    pub saved_at: UtcDateTime,
    pub route: Route,
    pub trip_state: Option<TripState>,
}
