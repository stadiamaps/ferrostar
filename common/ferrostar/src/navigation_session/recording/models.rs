use chrono::Utc;
use serde::{Deserialize, Serialize};

use crate::{
    models::Route,
    navigation_controller::{
        models::{
            NavigationControllerConfig, SerializableNavState,
            SerializableNavigationControllerConfig, TripState,
        },
        step_advance::SerializableStepAdvanceCondition,
    },
};

#[cfg(feature = "wasm-bindgen")]
use tsify::Tsify;

/// A builder for serializing a navigation recording.
#[derive(Serialize, Deserialize, Clone)]
pub struct NavigationRecordingMetadata {
    /// Version of Ferrostar used.
    pub version: String,
    /// Initial timestamp of the recording.
    pub initial_timestamp: i64,
    /// Configuration of the navigation controller.
    pub config: SerializableNavigationControllerConfig,
    /// Initial route used during navigation.
    pub initial_route: Route,
}

impl NavigationRecordingMetadata {
    /// Creates a new navigation recorder with route configuration and initial state.
    pub fn new(config: NavigationControllerConfig, initial_route: Route) -> Self {
        Self {
            version: env!("CARGO_PKG_VERSION").to_string(),
            initial_timestamp: Utc::now().timestamp_millis(),
            config: SerializableNavigationControllerConfig::from(config),
            initial_route,
        }
    }

    /// Serializes the navigation recording to a JSON string.
    pub fn to_json(&self, events: Vec<NavigationRecordingEvent>) -> Result<String, RecordingError> {
        let recording = NavigationRecording {
            recording: self.clone(),
            events,
        };
        serde_json::to_string(&recording).map_err(|e| RecordingError::SerializationError {
            error: e.to_string(),
        })
    }
}

/// A navigation session recording.
///
/// Internally this contains the full event stream.
// TODO: Hints for how you would typically use / interact with this? You can link to other types (and functions) by the way like [`NavigationReplay`] :)  and a list of navigation events.
#[derive(Serialize, Deserialize, Clone)]
pub struct NavigationRecording {
    #[serde(flatten)]
    pub recording: NavigationRecordingMetadata,
    /// List of navigation events in order of occurrence.
    pub events: Vec<NavigationRecordingEvent>,
}

impl NavigationRecording {
    /// Deserializes a previously saved navigation recording from a JSON string.
    pub fn from_json(json: &str) -> Self {
        serde_json::from_str(json)
            .map_err(|e| RecordingError::SerializationError {
                error: e.to_string(),
            })
            .unwrap()
    }
}

/// A session recording error.
#[derive(Debug, Serialize)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
pub enum RecordingError {
    /// Error during serialization.
    #[cfg_attr(
        feature = "std",
        error("Error serializing navigation recording: {error}.")
    )]
    SerializationError { error: String },

    /// Recording is not enabled for this controller.
    #[cfg_attr(
        feature = "std",
        error("Recording is not enabled for this controller.")
    )]
    RecordingNotEnabled,
}

/// An event that occurs during navigation.
///
/// This is used for the optional session recording / telemetry.
#[derive(Clone, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
pub struct NavigationRecordingEvent {
    /// The timestamp of the event in milliseconds since Jan 1, 1970 UTC.
    pub timestamp: i64,
    /// Data associated with the event.
    pub event_data: NavigationRecordingEventData,
}

impl NavigationRecordingEvent {
    pub fn new(event_data: NavigationRecordingEventData) -> Self {
        Self {
            timestamp: Utc::now().timestamp_millis(),
            event_data,
        }
    }

    /// Create a [`NavigationRecordingEventData::StateUpdate`] event from a [`SerializableNavState`]
    pub fn state_update(serializable_nav_state: SerializableNavState) -> Self {
        Self::new(NavigationRecordingEventData::StateUpdate {
            trip_state: serializable_nav_state.trip_state,
            step_advance_condition: serializable_nav_state.step_advance_condition,
        })
    }

    pub fn timestamp(&self) -> i64 {
        self.timestamp
    }
}

/// The event type.
///
/// For full replayability, we record things like rerouting, and not just location updates.
#[derive(Clone, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
pub enum NavigationRecordingEventData {
    StateUpdate {
        trip_state: TripState,
        step_advance_condition: SerializableStepAdvanceCondition,
    },
    // TODO: Figure out how to record re-routes.
    RouteUpdate {
        /// Updated route.
        route: Route,
    },
}
