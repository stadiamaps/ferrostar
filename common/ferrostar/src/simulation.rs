//! Tools for simulating progress along a route.
//!
//! # Example
//!
//! Here's an example usage with the polyline constructor.
//! This can serve as a template for writing your own test code.
//! You may also get some inspiration from the [Swift](https://github.com/stadiamaps/ferrostar/blob/main/apple/Sources/FerrostarCore/Location.swift)
//! or [Kotlin](https://github.com/stadiamaps/ferrostar/blob/main/android/core/src/main/java/com/stadiamaps/ferrostar/core/Location.kt)
//! `SimulatedLocationProvider` implementations which wrap this.
//!
//! ```
//! use ferrostar::simulation::{advance_location_simulation, location_simulation_from_polyline, LocationBias};
//! # use std::error::Error;
//! # fn main() -> Result<(), Box<dyn Error>> {
//!
//! let polyline_precision = 6;
//! // Build the initial state from an encoded polyline.
//! // You can create a simulation from coordinates or even a [Route] as well.
//! let mut state = location_simulation_from_polyline(
//!     "wzvmrBxalf|GcCrX}A|Nu@jI}@pMkBtZ{@x^_Afj@Inn@`@veB",
//!     polyline_precision,
//!     // Passing `Some(number)` will resample your polyline at uniform distances.
//!     // This is often desirable to create a smooth simulated movement when you don't have a GPS trace.
//!     None,
//!     LocationBias::None,
//! )?;
//!
//! loop {
//!     let mut new_state = advance_location_simulation(&state);
//!     if new_state == state {
//!         // When the simulation reaches the end, it keeps yielding the input state.
//!         break;
//!     }
//!     state = new_state;
//!     // Do something; maybe sleep for some period of time until the next timestamp?
//! }
//! #
//! # Ok(())
//! # }
//! ```

use crate::algorithms::trunc_float;
use crate::models::{CourseOverGround, GeographicCoordinate, Route, UserLocation};
use geo::{coord, Bearing, Densify, Geodesic, Haversine, LineString, Point};
use polyline::decode_polyline;

#[cfg(any(test, feature = "wasm-bindgen"))]
use serde::{Deserialize, Serialize};

#[cfg(feature = "wasm-bindgen")]
use wasm_bindgen::{prelude::*, JsValue};

#[cfg(feature = "wasm-bindgen")]
use tsify::Tsify;

#[cfg(all(feature = "std", not(feature = "web-time")))]
use std::time::SystemTime;

#[cfg(feature = "web-time")]
use web_time::SystemTime;

#[cfg(feature = "alloc")]
use alloc::{
    string::{String, ToString},
    vec::Vec,
};

#[derive(Debug)]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "wasm-bindgen", derive(Serialize, Deserialize, Tsify))]
#[cfg_attr(feature = "wasm-bindgen", tsify(into_wasm_abi, from_wasm_abi))]
pub enum SimulationError {
    #[cfg_attr(feature = "std", error("Failed to parse polyline: {error}."))]
    /// Errors decoding the polyline string.
    PolylineError { error: String },
    #[cfg_attr(feature = "std", error("Not enough points (expected at least two)."))]
    /// Not enough points in the input.
    NotEnoughPoints,
}

/// Controls how simulated locations deviate from the actual route line.
/// This simulates real-world GPS behavior where readings often have systematic bias.
#[derive(Clone, PartialEq, Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(any(feature = "wasm-bindgen", test), derive(Serialize, Deserialize))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
#[cfg_attr(feature = "wasm-bindgen", tsify(into_wasm_abi, from_wasm_abi))]
pub enum LocationBias {
    /// Simulates GPS bias by offsetting locations to the left of the route direction.
    /// The f64 parameter specifies the offset distance in meters.
    Left(f64),

    /// Simulates GPS bias by offsetting locations to the right of the route direction.
    /// The f64 parameter specifies the offset distance in meters.
    Right(f64),

    /// Simulates GPS bias by randomly choosing left or right offset on initialization
    /// and maintaining that bias throughout the route.
    /// The f64 parameter specifies the offset distance in meters.
    ///
    /// This mimics real-world GPS behavior where bias direction is random but typically
    /// remains consistent during a trip.
    Random(f64),

    /// No position bias - locations follow the route line exactly.
    ///
    /// This provides "perfect" GPS behavior, useful for testing basic route following
    /// without position uncertainty.
    None,
}

/// The current state of the simulation.
#[derive(Clone, PartialEq)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[cfg_attr(any(feature = "wasm-bindgen", test), derive(Serialize, Deserialize))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
#[cfg_attr(feature = "wasm-bindgen", tsify(into_wasm_abi, from_wasm_abi))]
pub struct LocationSimulationState {
    pub current_location: UserLocation,
    remaining_locations: Vec<GeographicCoordinate>,
    bias: LocationBias,
}

/// Creates a location simulation from a set of coordinates.
///
/// Optionally resamples the input line so that there is a maximum distance between points.
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn location_simulation_from_coordinates(
    coordinates: &[GeographicCoordinate],
    resample_distance: Option<f64>,
    bias: LocationBias,
) -> Result<LocationSimulationState, SimulationError> {
    if let Some((current, rest)) = coordinates.split_first() {
        if let Some(next) = rest.first() {
            let (jittered_current, bearing) = add_lateral_offset(*current, *next, &bias);

            let accuracy = match bias {
                LocationBias::None => 0.0,
                LocationBias::Left(m) | LocationBias::Right(m) | LocationBias::Random(m) => m,
            };

            let current_location = UserLocation {
                coordinates: jittered_current,
                horizontal_accuracy: accuracy,
                course_over_ground: Some(CourseOverGround::new(bearing, Some(5))),
                timestamp: SystemTime::now(),
                speed: None,
            };
            let remaining_locations = if let Some(distance) = resample_distance {
                // Interpolate so that there are no points further apart than the resample distance.
                let coords: Vec<_> = rest
                    .iter()
                    .map(|coord| {
                        coord! {
                            x: coord.lng,
                            y: coord.lat
                        }
                    })
                    .collect();
                let linestring: LineString = coords.into();
                let densified_linestring = linestring.densify::<Haversine>(distance);
                densified_linestring
                    .points()
                    .map(|point| GeographicCoordinate {
                        // We truncate the value to 6 digits of precision
                        // in line with standard navigation API practice.
                        // Nobody needs precision beyond this point,
                        // and it makes testing very annoying.
                        lat: trunc_float(point.y(), 6),
                        lng: trunc_float(point.x(), 6),
                    })
                    .collect()
            } else {
                Vec::from(rest)
            };

            Ok(LocationSimulationState {
                current_location,
                remaining_locations,
                bias,
            })
        } else {
            Err(SimulationError::NotEnoughPoints)
        }
    } else {
        Err(SimulationError::NotEnoughPoints)
    }
}

/// Creates a location simulation from a route.
///
/// Optionally resamples the route geometry so that there is no more than the specified maximum distance between points.
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn location_simulation_from_route(
    route: &Route,
    resample_distance: Option<f64>,
    bias: LocationBias,
) -> Result<LocationSimulationState, SimulationError> {
    // This function is purely a convenience for now,
    // but we eventually expand the simulation to be aware of route timing
    location_simulation_from_coordinates(&route.geometry, resample_distance, bias)
}

/// Creates a location simulation from a polyline.
///
/// Optionally resamples the input line so that there is no more than the specified maximum distance between points.
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn location_simulation_from_polyline(
    polyline: &str,
    precision: u32,
    resample_distance: Option<f64>,
    bias: LocationBias,
) -> Result<LocationSimulationState, SimulationError> {
    let linestring =
        decode_polyline(polyline, precision).map_err(|error| SimulationError::PolylineError {
            error: error.to_string(),
        })?;
    let coordinates: Vec<_> = linestring
        .coords()
        .map(|c| GeographicCoordinate::from(*c))
        .collect();
    location_simulation_from_coordinates(&coordinates, resample_distance, bias)
}

fn add_lateral_offset(
    current: GeographicCoordinate,
    next: GeographicCoordinate,
    bias: &LocationBias,
) -> (GeographicCoordinate, f64) {
    let current_point = Point::from(current);
    let next_point = Point::from(next);
    let bearing = Geodesic::bearing(current_point, next_point);

    match bias {
        LocationBias::None => (current, bearing),
        LocationBias::Left(meters) | LocationBias::Right(meters) | LocationBias::Random(meters) => {
            let sign = match bias {
                LocationBias::Left(_) => -1.0,
                LocationBias::Right(_) => 1.0,
                LocationBias::Random(_) => {
                    if rand::random() {
                        1.0
                    } else {
                        -1.0
                    }
                }
                LocationBias::None => unreachable!(),
            };

            // calculate perpendicular bearing (±90° from bearing)
            let lateral_bearing_rad = (bearing + sign * 90.0).to_radians();

            // offset to approximate degrees
            let offset_deg = meters / 111111.0;

            let lat_offset = offset_deg * lateral_bearing_rad.cos();
            let lng_offset = offset_deg * lateral_bearing_rad.sin();

            (
                GeographicCoordinate {
                    lat: current.lat + lat_offset,
                    lng: current.lng + lng_offset,
                },
                bearing,
            )
        }
    }
}

/// Returns the next simulation state based on the desired strategy.
/// Results of this can be thought of like a stream from a generator function.
///
/// This function is intended to be called once/second.
/// However, the caller may vary speed to purposefully replay at a faster rate
/// (ex: calling 3x per second will be a triple speed simulation).
///
/// When there are now more locations to visit, returns the same state forever.
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn advance_location_simulation(state: &LocationSimulationState) -> LocationSimulationState {
    if let Some((next_coordinate, rest)) = state.remaining_locations.split_first() {
        let (jittered_next, bearing) = add_lateral_offset(
            *next_coordinate,
            if let Some(future) = rest.first() {
                *future
            } else {
                *next_coordinate
            },
            &state.bias,
        );

        let accuracy = match state.bias {
            LocationBias::None => 0.0,
            LocationBias::Left(m) | LocationBias::Right(m) | LocationBias::Random(m) => m,
        };

        let next_location = UserLocation {
            coordinates: jittered_next,
            horizontal_accuracy: accuracy,
            course_over_ground: Some(CourseOverGround::new(bearing, Some(5))),
            timestamp: SystemTime::now(),
            speed: None,
        };

        LocationSimulationState {
            current_location: next_location,
            remaining_locations: Vec::from(rest),
            bias: state.bias.clone(),
        }
    } else {
        state.clone()
    }
}

/// JavaScript wrapper for `location_simulation_from_coordinates`.
#[cfg(feature = "wasm-bindgen")]
#[wasm_bindgen(js_name = locationSimulationFromCoordinates)]
pub fn js_location_simulation_from_coordinates(
    coordinates: JsValue,
    resample_distance: Option<f64>,
    bias: LocationBias,
) -> Result<JsValue, JsValue> {
    let coordinates: Vec<GeographicCoordinate> = serde_wasm_bindgen::from_value(coordinates)
        .map_err(|error| JsValue::from_str(&error.to_string()))?;

    location_simulation_from_coordinates(&coordinates, resample_distance, bias)
        .map(|state| serde_wasm_bindgen::to_value(&state).unwrap())
        .map_err(|error| JsValue::from_str(&error.to_string()))
}

/// JavaScript wrapper for `location_simulation_from_route`.
#[cfg(feature = "wasm-bindgen")]
#[wasm_bindgen(js_name = locationSimulationFromRoute)]
pub fn js_location_simulation_from_route(
    route: JsValue,
    resample_distance: Option<f64>,
    bias: LocationBias,
) -> Result<JsValue, JsValue> {
    let route: Route = serde_wasm_bindgen::from_value(route)
        .map_err(|error| JsValue::from_str(&error.to_string()))?;

    location_simulation_from_route(&route, resample_distance, bias)
        .map(|state| serde_wasm_bindgen::to_value(&state).unwrap())
        .map_err(|error| JsValue::from_str(&error.to_string()))
}

/// JavaScript wrapper for `location_simulation_from_polyline`.
#[cfg(feature = "wasm-bindgen")]
#[wasm_bindgen(js_name = locationSimulationFromPolyline)]
pub fn js_location_simulation_from_polyline(
    polyline: &str,
    precision: u32,
    resample_distance: Option<f64>,
    bias: LocationBias,
) -> Result<JsValue, JsValue> {
    location_simulation_from_polyline(polyline, precision, resample_distance, bias)
        .map(|state| serde_wasm_bindgen::to_value(&state).unwrap())
        .map_err(|error| JsValue::from_str(&error.to_string()))
}

/// JavaScript wrapper for `advance_location_simulation`.
#[cfg(feature = "wasm-bindgen")]
#[wasm_bindgen(js_name = advanceLocationSimulation)]
pub fn js_advance_location_simulation(state: JsValue) -> JsValue {
    let state: LocationSimulationState = serde_wasm_bindgen::from_value(state).unwrap();
    let new_state = advance_location_simulation(&state);
    serde_wasm_bindgen::to_value(&new_state).unwrap()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::algorithms::snap_user_location_to_line;
    use geo::{Distance, Haversine};
    use rstest::rstest;

    #[rstest]
    #[case(None)]
    #[case(Some(10.0))]
    fn advance_to_next_location(#[case] resample_distance: Option<f64>) {
        let mut state = location_simulation_from_coordinates(
            &[
                GeographicCoordinate { lng: 0.0, lat: 0.0 },
                GeographicCoordinate {
                    lng: 0.0001,
                    lat: 0.0001,
                },
                GeographicCoordinate {
                    lng: 0.0002,
                    lat: 0.0002,
                },
                GeographicCoordinate {
                    lng: 0.0003,
                    lat: 0.0003,
                },
            ],
            resample_distance,
            LocationBias::None,
        )
        .expect("Unable to initialize simulation");

        // Loop until state no longer changes
        let mut states = vec![state.clone()];
        loop {
            let new_state = advance_location_simulation(&state);
            if new_state == state {
                break;
            }
            state = new_state;
            states.push(state.clone());
        }

        insta::assert_yaml_snapshot!(format!("{:?}", resample_distance), states);
    }

    #[test]
    fn state_from_polyline() {
        let state = location_simulation_from_polyline(
            "wzvmrBxalf|GcCrX}A|Nu@jI}@pMkBtZ{@x^_Afj@Inn@`@veB",
            6,
            None,
            LocationBias::None,
        )
        .expect("Unable to parse polyline");
        insta::assert_yaml_snapshot!(state);
    }

    #[test]
    fn test_extended_interpolation_simulation() {
        let polyline = r#"umrefAzifwgF?yJf@?|C@?sJ?iL@_BBqD@cDzh@L|@?jBuDjCCl@u@^f@nB?|ABd@s@r@_AAiBBiC@kAlAHrEQ|F@pCNpA?pAAfB?~CkAtXsGRXlDw@rCo@jBc@SwAKoDr@}GLyAJ}AEs@]qBs@gE_@qC?aBBqAVkBZwBLmAFcBG_DOuB?}A^wAjA}Av@eBJoAAyA[sBbCUhAEIoCdAaCd@{@Fer@@ae@?aD?o[Ny@Vk@Sg@C_FCcDT[S_@Ow@F}oCXoAVe@_@e@?mE?cDNm@Og@Ok@Ck^N_BRu@a@OJqFFyDV[a@kAIkSLcF|AgNb@{@U_@JaEN}ETW[cA\_TbAkm@P_H\sE`AgFrCkKlAuGrEo\n@_B|@[~sBa@pAc@|AAh`Aa@jGEnGCrh@AfiAAjAx@TW`DO|CK\mEZ?~LBzBA|_@GtA?zPGlKQ?op@?uO@ggA?wE@uFEwXEyOCeFAkMAsKIot@?_FEoYAsI?yC?eH?}C?}GAy]Bux@Aog@AmKCmFC}YA}WVgBRu@vAaBlC{CxDCR?h@AhHQvGApDA|BAhHA`DC|GGzFDlM@jNA|J?bAkBtACvAArCClINfDdAfFGzW[|HI`FE@eMhHEt^KpJE"#;
        let max_distance = 10.0;
        let mut state =
            location_simulation_from_polyline(polyline, 6, Some(max_distance), LocationBias::None)
                .expect("Unable to create initial state");
        let original_linestring = decode_polyline(polyline, 6).expect("Unable to decode polyline");

        // Loop until state no longer changes
        let mut states = vec![state.clone()];
        loop {
            let new_state = advance_location_simulation(&state);
            if new_state == state {
                break;
            }

            // The distance between each point in the simulation should be <= max_distance
            let current_point: Point = state.current_location.into();
            let next_point: Point = new_state.current_location.into();
            let distance = Haversine::distance(current_point, next_point);
            // I'm actually not 100% sure why this extra fudge is needed, but it's not a concern for today.
            assert!(
                distance <= max_distance + 7.0,
                "Expected consecutive points to be <= {max_distance}m apart; was {distance}m"
            );

            let snapped =
                snap_user_location_to_line(new_state.current_location, &original_linestring);
            let snapped_point: Point = snapped.coordinates.into();
            let distance = Haversine::distance(next_point, snapped_point);
            assert!(
                distance <= max_distance,
                "Expected snapped point to be on the line; was {distance}m away"
            );

            state = new_state;
            states.push(state.clone());
        }

        // Sanity check: the simulation finishes on the last point
        assert_eq!(
            state.current_location.coordinates,
            original_linestring
                .points()
                .last()
                .expect("Expected at least one point")
                .into()
        );
        insta::assert_yaml_snapshot!(states);
    }

    #[rstest]
    #[case(LocationBias::None)]
    #[case(LocationBias::Left(4.0))]
    #[case(LocationBias::Right(4.0))]
    #[case(LocationBias::Random(4.0))]
    fn test_location_bias(#[case] bias: LocationBias) {
        let coordinates = vec![
            GeographicCoordinate { lng: 0.0, lat: 0.0 },
            GeographicCoordinate {
                lng: 0.0001,
                lat: 0.0001,
            },
            GeographicCoordinate {
                lng: 0.0002,
                lat: 0.0002,
            },
        ];

        let state = location_simulation_from_coordinates(&coordinates, None, bias.clone())
            .expect("Failed to create simulation");

        if matches!(bias, LocationBias::None) {
            assert_eq!(state.current_location.coordinates, coordinates[0]);
            return;
        }

        let expected_meters = match bias {
            LocationBias::Left(m) | LocationBias::Right(m) | LocationBias::Random(m) => m,
            LocationBias::None => unreachable!(),
        };

        let original_point: Point = coordinates[0].into();
        let offset_point: Point = state.current_location.coordinates.into();
        let distance = Haversine::distance(original_point, offset_point);

        assert!(
            (distance - expected_meters).abs() < 0.1,
            "Expected offset of {expected_meters}m but got {distance}m"
        );
    }

    #[test]
    fn test_bias_consistency() {
        let coordinates = vec![
            GeographicCoordinate { lng: 0.0, lat: 0.0 },
            GeographicCoordinate {
                lng: 0.0001,
                lat: 0.0001,
            },
            GeographicCoordinate {
                lng: 0.0002,
                lat: 0.0002,
            },
            GeographicCoordinate {
                lng: 0.0003,
                lat: 0.0003,
            },
        ];

        let mut state =
            location_simulation_from_coordinates(&coordinates, None, LocationBias::Random(4.0))
                .expect("Failed to create simulation");

        let first_point: Point = state.current_location.coordinates.into();
        let first_original: Point = coordinates[0].into();
        let initial_distance = Haversine::distance(first_point, first_original);

        while let Some((next, _)) = state.remaining_locations.split_first() {
            let new_state = advance_location_simulation(&state);
            if new_state == state {
                break;
            }

            let current_point: Point = new_state.current_location.coordinates.into();
            let original_point: Point = (*next).into();
            let distance = Haversine::distance(current_point, original_point);

            assert!(
                (distance - initial_distance).abs() < 0.1,
                "Bias distance changed from {initial_distance}m to {distance}m"
            );

            state = new_state;
        }
    }
}
