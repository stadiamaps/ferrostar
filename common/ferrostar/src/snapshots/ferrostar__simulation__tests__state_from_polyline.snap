---
source: ferrostar/src/simulation.rs
assertion_line: 417
expression: state
---
current_location:
  coordinates:
    lat: 60.534716
    lng: -149.543469
  horizontal_accuracy: 0
  course_over_ground:
    degrees: 288
    accuracy: 5
  speed: ~
remaining_locations:
  - lat: 60.534782
    lng: -149.543879
  - lat: 60.534829
    lng: -149.544134
  - lat: 60.534856
    lng: -149.5443
  - lat: 60.534887
    lng: -149.544533
  - lat: 60.534941
    lng: -149.544976
  - lat: 60.534971
    lng: -149.545485
  - lat: 60.535003
    lng: -149.546177
  - lat: 60.535008
    lng: -149.546937
  - lat: 60.534991
    lng: -149.548581
bias: None
